### PR TITLE
Issue #13155: False Negative: VariableDeclarationUsageDistance

### LINKED ISSUE #13155: update documentation: VariableDeclarationUsageDistance on  block of initialization methods
I have read check documentation: https://checkstyle.sourceforge.io/config_coding.html#VariableDeclarationUsageDistance
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
/var/tmp $ javac Test.java

/var/tmp $ cat config.xml
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.2//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_2.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="VariableDeclarationUsageDistance">
            <property name="allowedDistance" value="1"/>
        </module>
    </module>
</module>

/var/tmp $ cat Test.java
public class Test {
    public Test test(Object a) {
        Test obj = test("abc");
        int var = 12; // violation expected 'Dist.* between variable 'var'.* and its first usage is 4'
        obj.test("abc");
        obj.test("abc");
        obj.test("abc");
        int k = var + 3;
        return this;
    }

    Test fooFaaHooHaa() {
        int pc = 0; // violation expected, method chain is broken
        test("abc").test("abc");
        test("abc").test("abc");
        test("abc").test("abc");
        test("abc").test("abc");
        test("abc").test("abc");
        test("abc").test("abc");

        fooFaaHooHaa().test("abc");

        test(pc).test("abc");
        return null;
    }
}

/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-10.9.2-all.jar -c config.xml Test.java
Starting audit...
Audit done.
```

---

**Describe what you expect in detail.**
```bash
/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-10.9.x-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /home/vyom/IdeaProjects/Config-Checkstyle/src/Test.java:4:9: Distance between variable 'var' declaration and its first usage is 4, but allowed 1.  Consider making that variable final if you still need to store its value in advance (before method calls that might have side effects on the original value). [VariableDeclarationUsageDistance]
[ERROR] /home/vyom/IdeaProjects/Config-Checkstyle/src/Test.java:13:9: Distance between variable 'pc' declaration and its first usage is 8, but allowed 1.  Consider making that variable final if you still need to store its value in advance (before method calls that might have side effects on the original value). [VariableDeclarationUsageDistance]
Audit done.
Checkstyle ends with 2 errors.
```

There is unfinished work at https://github.com/checkstyle/checkstyle/pull/11672 and can be reused.

-------

Issue is fixed in some other update, no violations happening any more, but documentation should be updated to show example on how it works.

### PR DESCRIPTION
**Pull request by Kevin222004**
resolve #13155

Diff Regression config: https://gist.githubusercontent.com/Kevin222004/12b3bfed9dcea32830f532159c5023c6/raw/0618d0f5023f0d2c4100709cce815537cdbc509e/variable.xml
Report label: Report-3

### CODE REVIEW

**Comment by Kevin222004** [comment]
Github, generate report

**Comment by Kevin222004** [comment]
https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/c942fb6_2024101726/reports/diff/index.html

**Comment by Kevin222004** [comment]
Github, generate report

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/variabledeclarationusagedistance/InputVariableDeclarationUsageDistanceInitializationStatements.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+        void testMixedBranches2() {
+            int c = 12;
+            Test test = new Test();
+            test.test("abc");
+            test.test(c);
+        }
+
+        void testTypecast() {
+            int c = 12; // violation 'Dist.* between variable 'c'.* and its first usage is 5.'
```
Comment (Kevin222004) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/variabledeclarationusagedistance/InputVariableDeclarationUsageDistanceInitializationStatements.java:66]: This is not actually a violation.

Because is not returning correct name. 
https://github.com/checkstyle/checkstyle/blob/9842608983cf00c1927d5e5bfcffdb05f4936afb/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java#L226-L235
to fix this we actually have to update https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/api/FullIdent.java

I am fine to do that in this pr.

I have implemented some temporary code in file which is commented out right now which I am not sure it will work on every case or not. but the regression at https://github.com/checkstyle/checkstyle/pull/14479#issuecomment-1944500917 is generated with that code. and result look good to me.

**Comment by Kevin222004** [comment]
Github, generate report

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/variabledeclarationusagedistance/InputVariableDeclarationUsageDistanceInitializationStatements.java (hunk 1 above)
Comment (Kevin222004) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/variabledeclarationusagedistance/InputVariableDeclarationUsageDistanceInitializationStatements.java:66]: @nrmancuso any suggestion for this pr

Comment (Kevin222004) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/variabledeclarationusagedistance/InputVariableDeclarationUsageDistanceInitializationStatements.java:66]: @nrmancuso can you please start reviewing? I will find a way to reduce cyclomatic complexity.

Comment (nrmancuso) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/variabledeclarationusagedistance/InputVariableDeclarationUsageDistanceInitializationStatements.java:66]: @Kevin222004 lost track of this one. Looks like we have a new surviving mutation. Between dealing with that and the cyclomatic complexity, this code will probably change a lot, so I am not sure if it is valuable for me to review this yet.

Comment (Kevin222004) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/variabledeclarationusagedistance/InputVariableDeclarationUsageDistanceInitializationStatements.java:66]: I will resolve this cyclomatic complexity and pitest first.

Comment (Kevin222004) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/variabledeclarationusagedistance/InputVariableDeclarationUsageDistanceInitializationStatements.java:66]: @nrmancuso I think now it will be good to start reviewing. I will try to reply and update regularly.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java [hunk 2]
```diff
@@ -245,40 +246,83 @@ private static String getInstanceName(DetailAST methodCallAst) {
...
+    /**
+     * Checks if the provided instance names are equal.
+     *
+     * @param initInstanceName The init instance name.
+     * @param instanceName The instance name.
+     * @return true if the instance names is equal to initInstanceName.
+     */
+    private static boolean areInstanceNamesEqual(String initInstanceName, String instanceName) {
+        return !instanceName.isEmpty()
+                        && instanceName.equals(initInstanceName);
```
Comment (nrmancuso) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java:302]: What is the case where `initInstanceName` is empty?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java [hunk 3]
```diff
@@ -245,40 +246,83 @@ private static String getInstanceName(DetailAST methodCallAst) {
...
+                case TokenTypes.EXPR:
+                    if (currentSiblingAst.getFirstChild().getType() == TokenTypes.METHOD_CALL) {
+                        final DetailAST methodCallAst = currentSiblingAst.getFirstChild();
+                        final String instanceName = getInstanceName(methodCallAst);
+                        if (initInstanceName.isEmpty()) {
+                            initInstanceName = instanceName;
+                        }
+                        isVarNotInInitializationSequence = isVarNotInInitializationSequence
+                            && !isVariableUsed(variableName, methodCallAst);
+                        result = areInstanceNamesEqual(initInstanceName, instanceName);
```
Comment (nrmancuso) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java:268]: Can we extract this to a private method?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java [hunk 4]
```diff
@@ -245,40 +246,83 @@ private static String getInstanceName(DetailAST methodCallAst) {
...
-                final String currentVariableName =
-                        currentSiblingAst.findFirstToken(TokenTypes.IDENT).getText();
-                isUsedVariableDeclarationFound = variableName.equals(currentVariableName);
+            currentSiblingAst = currentSiblingAst.getPreviousSibling();
+        }
+        boolean ans = result;
+        if (variableUsageAst != null) {
+            ans = result && !isVarNotInInitializationSequence;
+        }
+        return ans;
```
Comment (nrmancuso) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java:290]: Please refactor this to not have mutable variables and use more descriptive names.

**Review by nrmancuso — CHANGES_REQUESTED** [review]
First pass:

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java (hunk 2 above)
Comment (Kevin222004) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java:302]: whenever in the sequence of the method first time that methodcall will be in method https://github.com/checkstyle/checkstyle/blob/3c6e287bb9d3c910e7173b0de5024ee94a0aabd4/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java#L249 at
```
                    if (currentSiblingAst.getFirstChild().getType() == TokenTypes.METHOD_CALL) {
                        final DetailAST methodCallAst = currentSiblingAst.getFirstChild();
                        final String instanceName = getInstanceName(methodCallAst);
                        if (initInstanceName.isEmpty()) {
                            initInstanceName = instanceName;
                        }
```

in this case initInstanceName would be empty.

Comment (Kevin222004) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java:302]: I assume you are asking about `!instanceName.isEmpty()`

```
public class InstanceNameEmpty {
    public void testIssue32_6() {
        Option aOpt = null; // violation 'Distance between .* declaration and its first usage is 3.'
        Option bOpt = null; // violation 'Distance between .* declaration and its first usage is 2.'
        Option cOpt = null;
        isNull(cOpt); // distance = 1
        isNull(bOpt); // distance = 2
        isNull(aOpt); // distance = 3
    }

    class Option {

        public void setArgName(String string) {
        }

    }

    boolean isNull(Option o) {
        return false;}

}
```

for this type of case, `isNull` method is defined within the same class. Therefore, it's not an instance method but within the same class. Therefore, the instance name for isNull will be empty.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java [hunk 5]
```diff
@@ -245,40 +246,83 @@ private static String getInstanceName(DetailAST methodCallAst) {
...
+    }
+
+    /**
+     * Checks if the specified variable name matches any identifier.
+     *
+     * @param variableName The name of the variable to search for within the AST.
+     * @param ast The root of the (AST) to search within.
+     * @return true if the variable name matches any identifier within the AST;
+     */
+    private static boolean isVariableUsed(String variableName, DetailAST ast) {
```
[the remark this answers is not in the document]
Comment (Kevin222004) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java:312]: I will update this in next push.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java (hunk 2 above)
Comment (nrmancuso) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java:302]: It seems like I didn't ask my question well; let me rephrase: how can we have an "empty" name? Conceptually, this doesn't make sense.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (Kevin222004) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java:312]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java (hunk 4 above)
Comment (Kevin222004) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java]: done 
I have make it simple.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/variabledeclarationusagedistance/InputVariableDeclarationUsageDistanceInitializationStatements.java (hunk 1 above)
Comment (Kevin222004) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/variabledeclarationusagedistance/InputVariableDeclarationUsageDistanceInitializationStatements.java:66]:
```
class Test {
    public Test test(Object a) {
        return this;
    }

    void testTypecast() {
        int c = 12; // violation 'Dist.* between variable 'c'.* and its first usage is 5.'
        Test test = new Test();
        test.test("abc");
        ((subclass)test).test("abc");
        test.test("abc");
        ((subclass)test).test("abc");
        test.test(c);
    }

    class subclass extends Test {
    }
}
```
in this code 
`((subclass)test).test("abc");`
The code casts the test object to its subclass (subclass) before calling the test method on it.
But The instance name is still test. The typecasting only changes the type of the reference, not the underlying object itself.
so whenever we call this method to get the instance name it will return `(` which is wrong we expect `test`
https://github.com/checkstyle/checkstyle/blob/4fe811819594747bcce6f62c216ea133c3cd2954/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java#L226-L235
and the violation is theirs.
so that means we have an issue with the method `createFullIdentBelow` in the `FullIdent` Class which needs to be figured out.

**Comment by Kevin222004** [comment]
@romani I am sorry I am quite busy with university task so trying to give time as possible. Will update by tomorrow.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java [hunk 6]
```diff
@@ -245,44 +246,99 @@ private static String getInstanceName(DetailAST methodCallAst) {
...
+     * Return the list of ident AST.
+     *
+     * @param ast The root of the (AST) to search within.
+     * @return list of ident ast within the given AST;
+     */
+    private static List<DetailAST> getIdentAst(DetailAST ast) {
+        DetailAST curNode = ast;
+        final List<DetailAST> identList = new ArrayList<>();
+        while (curNode.getType() != TokenTypes.SEMI) {
+            if (curNode.getType() == TokenTypes.IDENT) {
```
Comment (Kevin222004) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java:329]: A quick update on progression of this pr currently the ci of pitest is red because this is mutated 
```
Source File: "VariableDeclarationUsageDistanceCheck.java"
Class: "com.puppycrawl.tools.checkstyle.checks.coding.VariableDeclarationUsageDistanceCheck"
Method: "getIdentAst"
Line Contents: "if (curNode.getType() == TokenTypes.IDENT) {"
Mutator: "org.pitest.mutationtest.engine.gregor.mutators.RemoveConditionalMutator_EQUAL_IF"
Description: "removed conditional - replaced equality check with true"
Line Number: 329
```

I do have the test case to kill this the code like 
```
        void method() {
            Test non = new Test();
            String test = "check"; // violation ''
            non.test("b");
            non.test("b");
            non.test("a");
            non.test("a");
            non.test("a");
            non.test("a");
            non.test("a");
            System.out.println(test);
        }

```
Can kill this mutation

Comment (Kevin222004) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java:329]: but currently, we have again some other bug probably in calculating distance.
so that the code below is not showing the violation
```
kevin@kevin-Inspiron-15-5510:~/Desktop/check_style$ javac Test.java 
kevin@kevin-Inspiron-15-5510:~/Desktop/check_style$ 
kevin@kevin-Inspiron-15-5510:~/Desktop/check_style$ cat Test.java 
public class Test {

  public Test test(Object a) {
    return this;
  }

  void method() {
    Test non = new Test();
    String name = "Checkstyle";
    String test = "check"; // violation expected
    non.test("b");
    if (10 > 20) {
      System.out.println("maths rule changed");
    }
    if (20 > 10) {
      System.out.println("good");
    }
    System.out.println(name);
    System.out.println(test);
  }
  class subclass extends Test {
  }
}
kevin@kevin-Inspiron-15-5510:~/Desktop/check_style$ cat config.xml 
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.2//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_2.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="VariableDeclarationUsageDistance">
            <property name="allowedDistance" value="1"/>
        </module>
    </module>
</module>
kevin@kevin-Inspiron-15-5510:~/Desktop/check_style$ java -jar /home/kevin/Downloads/checkstyle-10.14.0-all.jar  -c config.xml Test.java 
Starting audit...
[ERROR] /home/kevin/Desktop/check_style/Test.java:9:5: Distance between variable 'name' declaration and its first usage is 4, but allowed 1.  Consider making that variable final if you still need to store its value in advance (before method calls that might have side effects on the original value). [VariableDeclarationUsageDistance]
Audit done.
Checkstyle ends with 1 errors.
kevin@kevin-Inspiron-15-5510:~/Desktop/check_style$ 
```
I expect a violation on line no 10.

I will figure out this ASAP and update this pr

**Comment by romani** [comment]
If no activity for one month, PR will be closed

**Comment by Kevin222004** [comment]
@romani for a few months till I get the job I cannot contribute.

**Comment by romani** [comment]
I am closing PR for now, we can reopen it later on.

Meanwhile everyone in web is welcome to continue this work to be done and be merged.
